### DISCUSSION TITLE
While looking at bash repo I noticed there is a method to install via zopen

### DISCUSSION THREAD

**Discussion by netskink** [zopencommunity/meta#168] [General]
Hello
I'm trying to get the meta package built in order to get make built, etc. for USS.

The first prerequisite packages use curl and pax.  The bash option also has same method but also mentions to `zopen` to install it.  What is `zopen`?  Is it the binary in the deprecated utils package?

**Comment by IgorTodorovskiIBM** [comment]
Hi! You may want to checkout https://github.com/ZOSOpenTools/meta/releases/tag/v1.0.0. This will get you bootstrapped with most of the tools.

After following the directions, you'll then have access to git and zopen. To get the latest zopen, you can clone the meta repo as follows:
```
git clone https://github.com/ZOSOpenTools/meta.git
cd meta
. ./.env
```

FYI, We recently changed `zopen download` to `zopen install` and zopen-setup may have the older version. 

Also more info on zopen here: https://zosopentools.github.io/meta/#/Guides/zopen

**Reply by netskink** [reply]
Thanks @IgorTodorovskiIBM.  So far, all I have done is pull the prerequisites down into a folder structure according to the meta/README.md.  That is:

```
          $HOME/zopen
                |
               +- boot
               |        |
               |       +- bash-5.2...zos
               |        |
               |       +- curl-7.03...zos
               |        |
               |       +- git-2.39...zos
               |        |
               |       +- make-4.4...zos
               |        |
               |       +-unzip-master-20230106...zos
               |
               +- prod
               |
              +-dev               
```

I looked at the tag notes for v0.0.1/v1.0.0 and noticed its 1 year+ old.   I'll continue with the README.md process since it's just 3 weeks old.  If that doesn't work, I'll fall back to the process in the tag you provided.  Thanks for that.


With that said, could you please enlighten me on the compiler.  I know the utils mentioned the open compiler and had a specifier for a v1.0.0 release.  I had the systems programmers install the v1.1 version for me.  However, the strings don't match.   Here is what I have access to.  


## The v2.4 and v2.4.1 compilers

Using the `--version` switch

```
$ /bin/xlc --help 2>&1 | grep V2.4.*C
  z/OS V2.4 XL C/C++

$ /bin/xlc++ --help 2>&1 | grep V2.4.*C
  z/OS V2.4 XL C/C++

$ /bin/xlclang --help 2>&1 | grep V2.4.*C
  z/OS V2.4.1 XL C/C++

$ /bin/xlclang++ --help 2>&1 | grep V2.4.*C
  z/OS V2.4.1 XL C/C++
```

Using the __COMPILER macro as done in deprecated utils/bin/lib/zopen-build

```
$ /bin/xlc -E -qshowmacros foo.cc | grep __COMPILER
#define __COMPILER_VER__ 0x42040000

$ /bin/xlc++ -E -qshowmacros foo.cc | grep __COMPILER
#define __COMPILER_VER__ 0x42040000

$ /bin/xlclang -E -qshowmacros foo.cc | grep __COMPILER
#define __COMPILER_VER__ 0x42030001

$ /bin/xlclang++ -E -qshowmacros foo.cc | grep __COMPILER
#define __COMPILER_VER__ 0x42030001
```

## The v1.1 open compiler

Using the `--version` switch

```
$ /usr/lpp/IBM/cnw/v1r1/openxl/bin/ibm-clang --version
IBM Open XL C/C++ 1.1 for z/OS, (5650-ZOS), clang version 14.0.0
Target: s390x-ibm-zos
Thread model: posix
InstalledDir: /usr/lpp/IBM/cnw/v1r1/openxl/bin

$ /usr/lpp/IBM/cnw/v1r1/openxl/bin/ibm-clang++ --version
IBM Open XL C/C++ 1.1 for z/OS, (5650-ZOS), clang version 14.0.0
Target: s390x-ibm-zos
Thread model: posix
InstalledDir: /usr/lpp/IBM/cnw/v1r1/openxl/bin

$ /usr/lpp/IBM/cnw/v1r1/openxl/bin/ibm-clang++64 --version
IBM Open XL C/C++ 1.1 for z/OS, (5650-ZOS), clang version 14.0.0
Target: s390x-ibm-zos
Thread model: posix
InstalledDir: /usr/lpp/IBM/cnw/v1r1/openxl/bin

$ /usr/lpp/IBM/cnw/v1r1/openxl/bin/ibm-clang64 --version
IBM Open XL C/C++ 1.1 for z/OS, (5650-ZOS), clang version 14.0.0
Target: s390x-ibm-zos
Thread model: posix
InstalledDir: /usr/lpp/IBM/cnw/v1r1/openxl/bin
```

Using the __COMPILER macro as done in deprecated utils/bin/lib/zopen-build

```

$ /usr/lpp/IBM/cnw/v1r1/openxl/bin/ibm-clang -E -m64 -dM foo.cc | grep __COMPILER
#define __COMPILER_VER__ 0x50000000

$ /usr/lpp/IBM/cnw/v1r1/openxl/bin/ibm-clang++ -E -m64 -dM foo.cc | grep __COMPILER
#define __COMPILER_VER__ 0x50000000

$ /usr/lpp/IBM/cnw/v1r1/openxl/bin/ibm-clang++64 -E -m64 -dM foo.cc | grep __COMPILER
#define __COMPILER_VER__ 0x50000000

$ /usr/lpp/IBM/cnw/v1r1/openxl/bin/ibm-clang64 -E -m64 -dM foo.cc | grep __COMPILER
#define __COMPILER_VER__ 0x50000000

```

However, I noted that when running the `$HOME/zopen/../unzip/bin/unzip -v` it reports 

```
$ ./unzip -v
...
Compiled with gcc 5655COE IBM C/C++ for Open Enterprise Languages on z/OS V1.0.0 for Unix on Jan  6 2023.
...
```

Pardon me, but I am just trying to make sure I have levelest on the compiler before I get too far down this endeavor.  Any guidance is appreciated.

**Reply by IgorTodorovskiIBM** [reply]
Ah, although the text in https://github.com/ZOSOpenTools/meta/releases/tag/v1.0.0 is old, the latest release is less than a month old. 
<img width="1313" alt="Screenshot 2023-02-09 at 10 35 53 AM" src="https://user-images.githubusercontent.com/39890068/217859225-71c5d4c7-7496-41e6-8234-c62935e55a7f.png">
But what you have so far looks fine.

We have not moved to the v1.1 compiler yet for our builds.  We are currently using this compiler: https://www-40.ibm.com/servers/resourcelink/svc00100.nsf/pages/xlCC++V241ForZOsV24?OpenDocument,  equivalent to IBM C/C++ for Open Enterprise Languages on z/OS V1.0.0 as delivered on z Modernization Stack. 

That would be `/bin/xlclang` for you. Let me know if that helps.

**Reply by netskink** [reply]
Hmm. I think we have that one.  See above where I posted the version strings.  However, when I attempted to build the old utils repo, it failed because of the __COMPILER define requirement. ie. it wasn't a match.  Consequently I asked for the v1.1 to be installed. 

Given this [comment](https://bugs.chromium.org/p/v8/issues/detail?id=13105#c5), it looks like you have a mix of compilers.

if you have some time @IgorTodorovskiIBM , is it possible to drop me an email via John.davis@broadcom.com?

**Comment by MikeFultonDev** [comment]
@netskink are you in good shape now or is there still work to do wrt compilers? If all good, can you close it off?

**Reply by netskink** [reply]
Yes this ok. Many thanks!

**Reply by netskink** [reply]
By the way it's a discussion thread, right? Not an issue, right?  Can it be closed?

**Reply by MikeFultonDev** [reply]
Yep. Close discussion is there. Or are you saying we should leave it open? If so then maybe we transfer it to the new ‘reference’ category I created?

**Comment by netskink** [comment]
Ahh, I see the close with comment.  Interesting.  Thanks for pointing that out!  Yes, its fine.


In summary, my understanding is that the compiler used for v8 differs from the compiler for the various z/OS Open Tools ports.  This makes sense.  v8 being newer uses clang.  Whereas the various ports are older and possibly in some cases predate clang and thus use a c compiler rather than a c++ compiler via clang.  Sorry if I am non percise.  I'm a compiler layman.  Please correct me if I am wrong.
